source/log.c

```diff
@@ -0,0 +1,302 @@
+
+#define AWS_CIRCULAR_DOUBLY_LIST_REMOVE(element) \
+    do { \
+        (element)->prev->next = (element)->next; \
+        (element)->next->prev = (element)->prev; \
+    } while (0)
+
+#define AWS_SINGLY_LIST_REVERSE(T, list) \
+    do { \
+        T* reverse_node__ = 0; \
+        while (list) { \
+            T* reverse_next__ = list->next; \
+            list->next = reverse_node__; \
+            reverse_node__ = list; \
+            list = reverse_next__; \
+        } \
+        list = reverse_node__; \
+    } while (0)
+
+struct aws_log_message {
+    struct aws_log_message *next;
+};
+
+struct aws_log_context {
+    struct aws_log_message *message_list;
+    struct aws_log_message *delete_list;
+    size_t max_message_len;
+    struct aws_memory_pool* message_pool;
+    struct aws_allocator *alloc;
+    struct aws_log_context *next;
+    struct aws_log_context *prev;
+};
+
+static void s_aws_log_default_report_function(const char* log_message) {
+    (void)log_message;
+}
+
+int s_log_thread_count;
+aws_log_report_fn* s_log_report_callback = s_aws_log_default_report_function;
+struct aws_mutex s_log_list_mutex = AWS_MUTEX_INIT;
+struct aws_log_context *s_log_list;
+
+AWS_THREAD_LOCAL struct aws_log_context *s_local_log_context;
+
+void aws_log_set_reporting_callback(aws_log_report_fn* report_callback) {
+    if (report_callback) {
+        s_log_report_callback = report_callback;
+    } else {
+        s_log_report_callback = s_aws_log_default_report_function;
+    }
+}
+
+int aws_vlog(enum aws_log_level level, const char *fmt, va_list va_args) {
+    (void)level;
+
+    if (!s_local_log_context) {
+        aws_raise_error(AWS_ERROR_LOG_UNINITIALIZED);
+        return AWS_OP_ERR;
+    }
+
+    /* Atomically grab and release all messages on the `delete_list` whenever present. */
+    struct aws_log_message* delete_list;
+    do {
+        delete_list = s_local_log_context->delete_list;
+    } while (!aws_atomic_cas_ptr((void **)&s_local_log_context->delete_list, delete_list, NULL));
+
+    while (delete_list) {
+        struct aws_log_message *next = delete_list->next;
+        aws_memory_pool_release(s_local_log_context->message_pool, delete_list);
+        delete_list = next;
+    }
+
+    /* Acquire memory for the new message. */
+    void *mem = aws_memory_pool_acquire(s_local_log_context->message_pool);
+    struct aws_log_message *msg = (struct aws_log_message *)mem;
+    char *msg_data = (char*)(msg + 1);
+    
+    /* Format the message. */
+    char date[256];
+    time_t now = time(NULL);
+    struct tm *t = localtime(&now);
+    strftime(date, sizeof(date) - 1, "%d %m %Y %H:%M", t);
+
+    char fmt_final[1024];
+    snprintf(fmt_final, sizeof(fmt_final), "[%s] %s [%d] %s\n", aws_log_level_to_string(level), date, (unsigned)aws_thread_current_thread_id(), fmt);
+
+    vsnprintf(msg_data, s_local_log_context->max_message_len, fmt_final, va_args);
+    if (strlen(msg_data) == s_local_log_context->max_message_len - 1) {

```

Comment:
You can make use of the return value of vsnprintf to avoid calling strlen, and to detect errors

Reply:
Looks like `vsnprintf` returns the number of characters that would be written if truncation didn't happen. I need to know if truncation happened here.

Reply:
if the number returned is the same size as your buffer then your buffer wasn’t big enough.

Reply:
Ah makes sense. I can get this soon.